gcwatch.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define _POSIX_C_SOURCE 199309L
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <jvmti.h>
+
+static int port;
+
+static bool inGc;
+static long start;
+
+static long getMillis()
+{
+   struct timespec ts;
+   clock_gettime(CLOCK_MONOTONIC, &ts);
+   return (ts.tv_sec * 1000) + (ts.tv_nsec / (1000 * 1000));
+}
+
+static void JNICALL gcStart(jvmtiEnv *jvmti)
+{
+   start = getMillis();
+   inGc = true;
+}
+
+static void JNICALL gcFinish(jvmtiEnv *jvmti)
+{
+   inGc = false;
```

Comment:
maybe reset start to zero, so it is symmetric  with the `gcStart` method?  Doesn't really matter.

Reply:
I guess that there could be a race condition between the call back  if we reset start to 0